firedrake/mesh.py

```diff
@@ -3162,57 +3180,40 @@ def _pic_swarm_in_mesh(
         tdim,
         gdim,
     )
-
-    # Set the SF graph for halos
-    owned_ranks_local_visible = owned_ranks_local[visible_idxs]
-    is_owned = parent_mesh.comm.rank == owned_ranks_local_visible
-    is_halo = ~is_owned
-    npts_owned = sum(is_owned)
-    npts_halo = sum(is_halo)
-    nroots = npts_owned + npts_halo  # roots should be all points
-    nleaves = npts_halo
-    local_halo_idxs = np.where(is_halo)[0].astype(IntType)  # what we pass as 'local' to the SF
-    remote_ranks = owned_ranks_local_visible[local_halo_idxs]
-
-    # Remote index is the offset in memory of the point on the other rank.
-    # Finding them is a bit of a faff...
-    remote_idxs = np.empty(npts_halo, dtype=IntType)
-    # global indices are unique particle identifiers - where they turn up on
-    # other ranks, they refer to the same particle.
-    global_idxs_local_visible = global_idxs_local[visible_idxs]
-    global_idxs_local_visible_halo = global_idxs_local_visible[local_halo_idxs]
-    global_idxs_local_visible_owned = global_idxs_local_visible[is_owned]
-    global_idxs_local_visible_owned_allranks = parent_mesh.comm.allgather(
-        global_idxs_local_visible_owned
-    )
-    # global_idxs_local_visible_owned_allranks is in rank order, so we can
-    # index into it with rank number (as long as our ranks start from zero!).
-    # We can now search for global index matches on each remote rank.
-    unique_remote_ranks = np.unique(remote_ranks)  # Don't need one for each halo point!
-    for unique_remote_rank in unique_remote_ranks:
-        remote_idxs[remote_ranks == unique_remote_rank] = np.where(
-            np.in1d(
-                global_idxs_local_visible_owned_allranks[unique_remote_rank],
-                global_idxs_local_visible_halo,
-            )
-        )[0]
-
-    # Interleave each rank and index into (rank, index) pairs for use as remote
-    # in the SF
-    remote_ranks_and_idxs = np.empty(2*nleaves, dtype=IntType)
-    remote_ranks_and_idxs[0::2] = remote_ranks
-    remote_ranks_and_idxs[1::2] = remote_idxs
-    sf = swarm.getPointSF()
-    sf.setGraph(nroots, local_halo_idxs, remote_ranks_and_idxs)
-    swarm.setPointSF(sf)
-
     # Note when getting original ordering for extruded meshes we recalculate
     # the base_parent_cell_nums and extrusion_heights - note this could
     # be an SF operation
     if redundant and parent_mesh.comm.rank != 0:
         original_ordering_swarm_coords = np.empty(shape=(0, coords.shape[1]))
     else:
         original_ordering_swarm_coords = coords
+    # Set pointSF
+    owners = swarm.getField("DMSwarm_rank")
+    swarm.restoreField("DMSwarm_rank")
+    halo_indices, = np.where(owners != parent_mesh.comm.rank)
+    halo_indices = halo_indices.astype(IntType)
+    n = coords.shape[0]
```

Comment:
Why does using the input coordinates at this point make sense? What if some of your coordinates are not found on ANY rank?

Reply:
`n` merely defines the local size of an array, i.e., `a`, that works as "broker". The set of indices of `a` is the target of `inputindex`; see `_make_input_ordering_sf`.
If some `index` of `a` corresponds to a missing point, `a[index]` is not updated when we `reduce` and it does not update any leaf data, i.e., `b`, when we `bcast`.

Reply:
Please make this clear in a comment

Reply:
Added.